 html.rb     |    2 +-
 issue.rhtml |    4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/html.rb b/html.rb
index 65aa597..aae84fa 100644
--- a/html.rb
+++ b/html.rb
@@ -39,7 +39,7 @@ class ErbHtml
   ###
 
   def h o; o.to_s.gsub("&", "&amp;").gsub("<", "&lt;").gsub(">", "&gt;") end
-  def t o; o.strftime "%Y-%m-%d %H:%M %Z" end
+  def t o; o.getlocal.strftime "%Y-%m-%d %H:%M %z" end
   def p o; "<p>" + h(o.to_s).gsub("\n\n", "</p><p>") + "</p>" end
   def obscured_email e; h e.gsub(/@.*?(>|$)/, "@...\\1") end
   def link_to o, name
diff --git a/issue.rhtml b/issue.rhtml
index a71a5e6..0220b88 100644
--- a/issue.rhtml
+++ b/issue.rhtml
@@ -18,7 +18,7 @@
 
 <% if issue.desc && !issue.desc.empty? %>
   <div class="description">
-  <%= link_issue_names project, p(issue.desc) %>
+  <pre><%= link_issue_names project, p(issue.desc) %></pre>
   </div>
 <% end %>
 
@@ -104,7 +104,7 @@
     </tr>
     <% unless comment.empty? %>
       <tr><td colspan="3" class="logcomment">
-      <%= link_issue_names project, comment %>
+      <pre><%= link_issue_names project, comment %></pre>
       </td></tr>
     <% end %>
     <tr><td></td></tr>
