 pass_keys.py | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git pass_keys.py pass_keys.py
index 0beda6c..830da51 100644
--- pass_keys.py
+++ pass_keys.py
@@ -29,11 +29,15 @@ def main():
     pass
 
 
+def is_direction(operation):
+    return re.search("^(top|bottom|left|right)$", operation, re.I)
+
+
 @result_handler(no_ui=True)
 def handle_result(args, result, target_window_id, boss):
     direction = args[1]
     key_mapping = args[2]
-    vim_id = args[3] if len(args) > 3 else "n?vim"
+    vim_id = args[3] if len(args) > 3 else "n?vim?"
 
     window = boss.window_id_map.get(target_window_id)
 
@@ -43,5 +47,7 @@ def handle_result(args, result, target_window_id, boss):
         for keymap in key_mapping.split(">"):
             encoded = encode_key_mapping(window, keymap)
             window.write_to_child(encoded)
-    else:
+    elif is_direction(direction):
         boss.active_tab.neighboring_window(direction)
+    else:
+        boss.active_tab.focus_visible_window()
